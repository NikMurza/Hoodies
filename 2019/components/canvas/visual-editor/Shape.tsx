import * as React from 'react';
import { mapTypeToComponent } from '@common/components/lib/canvas/shapes';
import {
  VisualEditorActions,
  VisualEditorContextMenuIdentity,
  VisualEditorShape,
  VisualEditorSize,
  VisualEditorComponentData,
} from '@common/components/lib/types/visual-editor';
import { typeContextMenuId } from '@common/components/lib/canvas/components/context-menu/with-context-menu/WithContextMenu';

interface Props {
  item: VisualEditorShape;
  contextMenuId: string;
  stageSize: VisualEditorSize;
  hoveredShape: { id: string; type: string };
  onShapeAction: <T extends VisualEditorContextMenuIdentity>(
    action: VisualEditorActions,
    data: T,
  ) => void;
  onShapeAddComponent: (data: VisualEditorComponentData) => void;
  onShapeRemoveComponent: (id: string) => void;
}

const Shape: React.FC<Props> = ({
  item,
  contextMenuId,
  stageSize,
  hoveredShape,
  onShapeAction,
  onShapeAddComponent,
  onShapeRemoveComponent,
}) => {
  const { id, type } = item;
  const ShapeComponent = mapTypeToComponent[type];
  const isContextMenuShow = contextMenuId === typeContextMenuId(id, type);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const props = item as any;

  // перехватываем onAction у shape
  const onAction = React.useCallback(
    <T extends VisualEditorContextMenuIdentity>(action: VisualEditorActions, data: T): void => {
      onShapeAction(action, data);
      if (props.onAction) props.onAction(action, data);
    },
    [props, onShapeAction],
  );
  return (
    <ShapeComponent
      key={`${type}_${id}`}
      {...props}
      onAction={onAction}
      isHovered={id === hoveredShape.id && type === hoveredShape.type}
      isContextMenuShow={isContextMenuShow}
      stageSize={stageSize}
      onAddComponent={onShapeAddComponent}
      onRemoveComponent={onShapeRemoveComponent}
    />
  );
};

export default Shape;
