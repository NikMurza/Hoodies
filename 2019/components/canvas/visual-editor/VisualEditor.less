.visual-editor {
  position: relative;
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;

  & .visual-canvas {
    position: relative;
    overflow: hidden;
    outline: none;

    &-scrollable {
      position: relative;
      flex-grow: 1;
      overflow: auto;
      &_hidden-scroll {
        -ms-overflow-style: none;
        scrollbar-width: none;
      }
      &_hidden-scroll::-webkit-scrollbar {
        display: none;
      }
    }
    &__components {
      position: absolute;
      top: 0;
      left: 0;
    }
    &__context-menu-wrapper {
      position: absolute;
    }
  }
}
