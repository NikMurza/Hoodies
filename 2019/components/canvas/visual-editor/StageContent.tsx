import * as React from 'react';
import { Store } from 'redux';
import { Provider } from 'react-redux';
import * as ReactKonva from 'react-konva';

import {
  VisualEditorActions,
  VisualEditorContextMenuIdentity,
  VisualEditorShape,
  VisualEditorSize,
  VisualEditorComponentData,
} from '@common/components/lib/types/visual-editor';
import Shape from './Shape';

export interface Props {
  store?: Store;
  items?: VisualEditorShape[];
  hoveredShape: { id: string; type: string };
  contextMenuId: string;
  stageSize: VisualEditorSize;
  onShapeAction: <T extends VisualEditorContextMenuIdentity>(
    action: VisualEditorActions,
    data: T,
  ) => void;
  onShapeAddComponent: (data: VisualEditorComponentData) => void;
  onShapeRemoveComponent: (id: string) => void;
}

const StageContent: React.FC<Props> = ({
  store,
  items,
  hoveredShape,
  contextMenuId,
  stageSize,
  onShapeAction,
  onShapeAddComponent,
  onShapeRemoveComponent,
}) => {
  const stage = React.useMemo(() => {
    const isHoveredShape = (shape: VisualEditorShape): boolean =>
      shape.id === hoveredShape.id && shape.type === hoveredShape.type;
    const ordered = items.filter((item) => !isHoveredShape(item));
    ordered.push(...items.filter(isHoveredShape));

    const renderItem = (item: VisualEditorShape): JSX.Element =>
      React.createElement(Shape, {
        key: `${item.type}-${item.id}`,
        item,
        contextMenuId,
        stageSize,
        hoveredShape,
        onShapeAction,
        onShapeAddComponent,
        onShapeRemoveComponent,
      });

    return <ReactKonva.Layer>{ordered.map(renderItem)}</ReactKonva.Layer>;
  }, [
    items,
    hoveredShape,
    contextMenuId,
    stageSize,
    onShapeAction,
    onShapeAddComponent,
    onShapeRemoveComponent,
  ]);

  return store ? <Provider store={store}>{stage}</Provider> : stage;
};

export default StageContent;
