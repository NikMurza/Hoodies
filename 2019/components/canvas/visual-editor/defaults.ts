import colorsConst from '@common/styles/constants/colors';
import fontConst from '@common/styles/constants/fonts';
import { VisualEditorSize } from '@common/components/lib/types/visual-editor';

interface ScaleDefaults {
  initial: number;
  min: number;
  max: number;
  step: number;
}
export interface CanvasDefaults {
  size: VisualEditorSize;
  gridCell: number;
  gridLineWidth: number;
  canvasOffset: number;
  zoom: ScaleDefaults;
}

// измерения холста
export const canvas: CanvasDefaults = {
  size: {
    width: window.innerWidth,
    height: window.innerHeight,
  },
  gridCell: 96, // сторона ячейки сетки
  gridLineWidth: 1,
  canvasOffset: 24, // минимально допустимое расстояние от края холста до элемента,
  zoom: {
    initial: 1,
    min: 0.1,
    max: 4,
    step: 0.1,
  },
};

// размер элемента по-умолчанию
export const element = {
  width: 200,
  height: 100,
  indent: 100, // размер отступа (по-умолчанию) между элементами
  radius: 60, // Радиус элемента (по-умолчанию)
  borderRadius: 0,
  borderWidth: 1,
  referencePointRadius: 6,
  editInputWidth: 100,
  editInputHeight: 24,
};

export const line = {
  borderWidth: 2,
  hoverWidth: 20,
  endpointLength: 11.5,
  endpointWidth: 10,
  iconRadius: 16,
};

export const font = {
  family: fontConst.family,
  textSize: 14,
  accentTextSize: 16,
  secondaryTextSize: 13,
};

export const colors = {
  backgroundColor: colorsConst.white,
  transparent: 'transparent',
  gridLinesColor: '#ECEDEF',
  shadowColor: '#cccccc',
  canvas: '#F2F3F4',
  circleColor: '#9DAFED',
  circleBackgroundColor: '#A469AD',
  circleBorderColor: '#8B5A94',
  circleActiveBackgroundColor: '#3B5EDC',
  circleActiveBorderColor: '#FFFFFF',
  subTextColor: '#93989E',
  sourceTypeColor: colorsConst.green,
  destTypeColor: '#59BDF8',
  convertorTypeColor: '#FFC700',
  blockShadowColor: 'rgba(115,115,115,0.12)',
  blockActiveColor: colorsConst.blue,
  borderColor: colorsConst.darkAdditional,
  referencePointsColor: colorsConst.white,
  referencePointsHoverColor: colorsConst.blue,
  referencePointsBorderColor: colorsConst.blue,
  iconSettingBackgroundColor: colorsConst.white,
  iconSettingColor: colorsConst.white,
};
