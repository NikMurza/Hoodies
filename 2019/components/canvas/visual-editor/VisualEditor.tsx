import * as React from 'react';

import { ReactReduxContext } from 'react-redux';
import Konva from 'konva';
import * as ReactKonva from 'react-konva';

import './VisualEditor.less';
import {
  VisualEditorActionCallback,
  VisualEditorActions,
  VisualEditorContextMenuIdentity,
  VisualEditorIdentityPosition,
  VisualEditorKeyboardAction,
  VisualEditorPosition,
  VisualEditorPositionalClick,
  VisualEditorShape,
  VisualEditorShapeWithRef,
  VisualEditorSize,
} from '@common/components/lib/types/visual-editor';
import { MouseButtons } from '@common/components/lib/types/mouseEvent';
import * as defaults from '@common/components/lib/canvas/visual-editor/defaults';
import { usePosition } from '@common/components/lib/canvas/hooks/usePosition';
import { useIsStage } from '@common/components/lib/canvas/hooks/useIsStage';
import WithDrop from '@common/components/lib/canvas/visual-editor/with-drop/WithDrop';
import WithResizeDetect, {
  WithResizeable,
} from '@common/components/lib/canvas/visual-editor/with-resize-detect/WithResizeDetect';
import StageContent from './StageContent';
import useAddRemoveComponents from './use-add-remove-component/useAddRemoveComponent';

export interface VELayer extends Konva.LayerConfig {
  children: JSX.Element;
}

export interface Props {
  /**
   * Идентификатор конкретного экземпляра VE
   */
  id?: string;

  /**
   * Элементы для отрисовки
   */
  items?: VisualEditorShape[];

  /*
   * Дополнительные слои ниже основного с контентом
   * Рекомендуется в них отображать дополнительные данные — grid, FPS и пр
   * См. useFPS, useGrid, useZoom
   */
  extraLayers?: JSX.Element;

  /*
   * Положение Stage
   * Значение по умолчанию - { x: 0, y: 0 }
   */
  position?: VisualEditorPosition;

  /*
   * Масштабирование содержимого Visual Editor, указывается в долях от 0,1 и выше.
   * Значение по умолчанию - 1, что соответствует 100%
   */
  zoom?: number;

  /*
   * Контролируем возможность перемещаться по канвасу
   * По умолчанию true
   */
  panEnabled?: boolean;

  /**
   * Вызывается на каких-либо событиях (изменение, удаление, контекстное меню)
   * @param action Событие
   * @param data Данные
   */
  onAction?: VisualEditorActionCallback;
}

const VisualEditorContainer = WithDrop(
  WithResizeDetect(
    (
      props: React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement> &
        WithResizeable,
    ): JSX.Element => {
      const { resizeableRef, ...rest } = props;
      return <div ref={resizeableRef} {...rest} />;
    },
  ),
);

/**
 * Визуальный редактор - отображает таблицы со связями (и не только)
 * @param props Свойства
 */
const VisualEditor: React.FC<React.PropsWithChildren<Props>> = ({
  id,
  position,
  zoom: customZoom = 1,
  items,
  extraLayers,
  children,
  panEnabled,
  onAction,
}) => {
  const [contextMenuId, setContextMenuId] = React.useState<string>('');
  const [hoveredShape, setHoveredShape] = React.useState({ id: '', type: '' });

  const [stageSize, setStageSize] = React.useState<VisualEditorSize>({
    width: window.innerWidth,
    height: window.innerHeight,
  });

  const stagePosition: VisualEditorPosition = {
    x: position?.x ?? 0,
    y: position?.y ?? 0,
  };

  const zoom = customZoom ?? defaults.canvas.zoom.initial;

  const stageRef = React.useRef<ReactKonva.Stage>(null);
  const isStage = useIsStage<Event>();

  const { components, onShapeAddComponent, onShapeRemoveComponent } = useAddRemoveComponents();

  const hideContextMenu = React.useCallback((): void => {
    setContextMenuId('');
  }, []);

  React.useEffect(() => {
    onAction<VisualEditorShapeWithRef<ReactKonva.Stage>>(VisualEditorActions.Mount, {
      forwardRef: stageRef,
    });
  }, [stageRef]);

  // Закрываем контекстное меню при правом клике на пустую область,
  // восстанавливаем действие по умолчанию.
  const onStageContextMenu = React.useCallback(
    (e: Konva.KonvaEventObject<PointerEvent>): void => {
      if (isStage(e)) {
        e.evt.preventDefault();
        const { x, y } = usePosition<PointerEvent>(e);

        onAction<VisualEditorPositionalClick>(VisualEditorActions.ContextMenu, {
          id,
          x,
          y,
          target: e.target,
          event: { ...e?.evt, button: MouseButtons.Right },
        });
      }
    },
    [hideContextMenu, onAction],
  );

  const onStageClick = React.useCallback(
    (e: Konva.KonvaEventObject<PointerEvent>): void => {
      const { x, y } = usePosition<PointerEvent>(e);
      onAction<VisualEditorPositionalClick>(VisualEditorActions.Click, {
        id,
        x,
        y,
        event: e?.evt,
        target: e?.target,
      });
      hideContextMenu();
    },
    [id, onAction],
  );

  const onStageMouseAction = React.useCallback(
    (action: VisualEditorActions) => (e: Konva.KonvaEventObject<MouseEvent>): void => {
      const { x, y } = usePosition<MouseEvent>(e);
      onAction<VisualEditorIdentityPosition>(action, {
        id,
        x,
        y,
      });
    },
    [id, onAction],
  );

  const onStageMouseMove = React.useCallback(onStageMouseAction(VisualEditorActions.MouseMove), [
    stageRef,
    onStageMouseAction,
  ]);

  const onStageMouseUp = React.useCallback(onStageMouseAction(VisualEditorActions.MouseUp), [
    stageRef,
    onStageMouseAction,
  ]);

  const onKeyboardAction = React.useCallback(
    (key: VisualEditorActions, event: React.KeyboardEvent<HTMLDivElement>): void => {
      onAction<VisualEditorKeyboardAction>(key, {
        id,
        isCtrl: event.ctrlKey,
        isShift: event.shiftKey,
        isAltKey: event.altKey,
        keyCode: event.keyCode,
        key: event.key,
        isRepeat: event.repeat,
      });
    },
    [id, onAction],
  );

  const onStageKeyDown = React.useCallback(
    (event: React.KeyboardEvent<HTMLDivElement>): void =>
      onKeyboardAction(VisualEditorActions.KeyDown, event),
    [onKeyboardAction],
  );

  const onStageKeyUp = React.useCallback(
    (event: React.KeyboardEvent<HTMLDivElement>): void =>
      onKeyboardAction(VisualEditorActions.KeyUp, event),
    [onKeyboardAction],
  );

  const onStageDrag = React.useCallback(
    (e: Konva.KonvaEventObject<DragEvent>) => {
      if (!isStage(e)) {
        return;
      }

      onAction<VisualEditorIdentityPosition>(VisualEditorActions.Move, {
        id,
        x: -e.target.x(),
        y: -e.target.y(),
      });
    },
    [id, onAction],
  );

  const onStageResize = React.useCallback(
    (size: VisualEditorSize): void => {
      setStageSize(size);
      onAction<VisualEditorSize>(VisualEditorActions.Resize, size);
    },
    [onAction],
  );

  const onShapeAction = React.useCallback(
    <T extends VisualEditorContextMenuIdentity>(action: VisualEditorActions, data: T): void => {
      const actions = {
        [VisualEditorActions.ContextMenu]: (data: VisualEditorContextMenuIdentity): void =>
          setContextMenuId(data.contextMenuId),
        [VisualEditorActions.Move]: hideContextMenu,
        [VisualEditorActions.Hover]: (data: VisualEditorShape): void => {
          setHoveredShape({ id: data.id, type: data.type });
        },
      };

      if (actions[action]) actions[action](data);
    },
    [hideContextMenu],
  );

  React.useEffect(() => {
    const selector = document.getElementById('root');
    if (!selector) {
      return null;
    }
    selector.addEventListener('click', hideContextMenu);
    selector.addEventListener('contextmenu', hideContextMenu, true);
    return (): void => {
      selector.removeEventListener('click', hideContextMenu);
      selector.removeEventListener('contextmenu', hideContextMenu);
    };
  }, []);

  const renderStage: boolean = stageSize.width > 0 && stageSize.height > 0;

  return (
    <VisualEditorContainer
      onResize={onStageResize}
      className="visual-editor"
      stageRef={stageRef}
      onAction={onAction}
    >
      <div
        tabIndex={-1}
        className="visual-canvas"
        onKeyDown={onStageKeyDown}
        onKeyUp={onStageKeyUp}
      >
        <ReactReduxContext.Consumer>
          {({ store }): JSX.Element =>
            !renderStage ? null : (
              <ReactKonva.Stage
                draggable={panEnabled}
                ref={stageRef}
                id={id}
                x={-stagePosition.x}
                y={-stagePosition.y}
                width={stageSize.width}
                height={stageSize.height}
                scaleX={zoom}
                scaleY={zoom}
                stageRef={stageRef}
                onClick={onStageClick}
                onContextMenu={onStageContextMenu}
                onMouseMove={onStageMouseMove}
                onMouseUp={onStageMouseUp}
                onDragEnd={onStageDrag}
              >
                {extraLayers}
                {/* подключаем к стору конву, если задан store */}
                {React.createElement(StageContent, {
                  store,
                  items,
                  hoveredShape,
                  contextMenuId,
                  stageSize,
                  onShapeAction,
                  onShapeAddComponent,
                  onShapeRemoveComponent,
                })}
              </ReactKonva.Stage>
            )
          }
        </ReactReduxContext.Consumer>
        {/* отрисовываем дополнительные компоненты */}
        <div className="visual-canvas__components">
          {Object.values(components)}
          {/* отрисовываем детей */}
          {children}
        </div>
      </div>
    </VisualEditorContainer>
  );
};

// по-умолчанию редактор занимает размер экрана
VisualEditor.defaultProps = {
  id: 'VisualEditor',
  items: [],
  panEnabled: true,
  onAction: (): void => undefined,
};

export default VisualEditor;
