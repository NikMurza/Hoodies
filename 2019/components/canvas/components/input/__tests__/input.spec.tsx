import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import * as defaults from '@common/components/lib/canvas/visual-editor/defaults';
import Input from '../input';

describe('Input component', () => {
  it('renders correctly', () => {
    const id = 'table1';
    const x = 100;
    const y = 100;
    const text = 'table1';
    const textColor = '#000';

    const props = {
      onComplete: jest.fn(),
      onCancel: jest.fn,
    };

    const onCompleteSpy = jest.spyOn(props, 'onComplete');
    const onCancelSpy = jest.spyOn(props, 'onCancel');

    const wrapper = shallow(
      <Input
        key={id}
        id={id}
        text={text}
        x={x}
        y={y}
        onComplete={props.onComplete}
        onCancel={props.onCancel}
        textColor={textColor}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();

    const input = wrapper.find('input');

    input.simulate('keydown', { key: 'Enter' });
    expect(onCompleteSpy).toBeCalledWith(text);

    input.simulate('keydown', { key: 'Escape' });
    expect(onCancelSpy).toBeCalled();
  });

  it('Input defaultProps', () => {
    expect(JSON.stringify(Input.defaultProps)).toEqual(
      JSON.stringify({
        width: defaults.element.editInputWidth,
        height: defaults.element.editInputHeight,
        onComplete: (): void => undefined,
        onCancel: (): void => undefined,
      }),
    );
  });
});
