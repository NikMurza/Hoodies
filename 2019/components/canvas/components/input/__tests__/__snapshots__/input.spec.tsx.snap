// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component renders correctly 1`] = `
<div
  className="edit-input"
  style={
    Object {
      "height": 24,
      "left": 50,
      "top": 88,
      "transform": "rotate(0deg)",
      "width": 100,
    }
  }
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="table1"
  />
</div>
`;
